ddtrace/internal/datadog/profiling/src/exporter.cpp

```diff
@@ -0,0 +1,560 @@
+  labels[cur_label].key = to_slice(key_sv);
+  labels[cur_label].num = val;
+  cur_label++;
+}
+
+void Profile::clear_buffers() {
+  locations.clear();
+  lines.clear();
+  std::fill(values.begin(), values.end(), 0);
+  std::fill(std::begin(labels), std::end(labels), ddog_prof_Label{});
+  cur_label = 0;
+  nframes = 0;
+}
+
+bool Profile::flush_sample() {
+  // We choose to normalize thread counts against the user's indicated
+  // preference, even though we have no control over how many frames are sent.
+  if (nframes > max_nframes) {
+    auto dropped_frames = nframes - max_nframes;
+    std::string name = "<" + std::to_string(dropped_frames) + " frame" +
+                       (1 == dropped_frames ? "" : "s") + " omitted>";
+    Profile::push_frame_impl(name, "", 0, 0);
+  }
+
+  ddog_prof_Sample sample = {
+      .locations = {&locations[0], locations.size()},
+      .values = {&values[0], values.size()},
+      .labels = {labels, cur_label},
+  };
+
+  ddog_prof_Profile_AddResult address =
+      ddog_prof_Profile_add(ddog_profile, sample);
+  if (address.tag == DDOG_PROF_PROFILE_ADD_RESULT_ERR) {
+    std::string ddog_errmsg(ddog_Error_message(&address.err).ptr);
+    errmsg = "Could not flush sample: " + errmsg;
+    ddog_Error_drop(&address.err);
+
+    clear_buffers();
+    return false;
+  }
+
+  clear_buffers();
+  return true;
+}
+
+bool Datadog::Profile::push_cputime(int64_t cputime, int64_t count) {
+  // NB all push-type operations return bool for semantic uniformity,
+  // even if they can't error.  This should promote generic code.
+  if (type_mask & ProfileType::CPU) {
+    values[val_idx.cpu_time] += cputime * count;
+    values[val_idx.cpu_count] += count;
+  }
+  return true;
+}
+
+bool Datadog::Profile::push_walltime(int64_t walltime, int64_t count) {
+  if (type_mask & ProfileType::Wall) {
+    values[val_idx.wall_time] += walltime * count;
+    values[val_idx.wall_count] += count;
+  }
+  return true;
+}
+
+bool Datadog::Profile::push_exceptioninfo(std::string_view exception_type,
+                                          int64_t count) {
+  if (type_mask & ProfileType::Exception) {
+    push_label(ExportLabelKey::exception_type, exception_type);
+    values[val_idx.exception_count] += count;
+  }
+  return true;
+}
+
+bool Datadog::Profile::push_acquire(int64_t acquire_time, int64_t count) {
+  if (type_mask & ProfileType::LockAcquire) {
+    values[val_idx.lock_acquire_time] += acquire_time;
+    values[val_idx.lock_acquire_count] += count;
+  }
+  return true;
+}
+
+bool Datadog::Profile::push_release(int64_t release_time, int64_t count) {
+  if (type_mask & ProfileType::LockRelease) {
+    values[val_idx.lock_release_time] += release_time;
+    values[val_idx.lock_release_count] += count;
+  }
+  return true;
+}
+
+bool Datadog::Profile::push_alloc(uint64_t size, uint64_t count) {
+  if (type_mask & ProfileType::Allocation) {
+    values[val_idx.alloc_space] += size;
+    values[val_idx.alloc_count] += count;
```

Comment:
minor: returning true when we are not pushing anything might seem strange.

Reply:
Very, very true.  Fixed in 61f10cc50, although I don't know whether I'll consume the return in the PR.

Reply:
Just realized that I inverted the logic--fixing now in case someone is looking at this and scratching their heads.